src/storage/txn/commands/flashback_to_version.rs

```diff
@@ -67,42 +71,50 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for FlashbackToVersion {
             context.statistics,
         );
         let mut txn = MvccTxn::new(TimeStamp::zero(), context.concurrency_manager);
-
-        let mut next_lock_key = self.next_lock_key.take();
-        let mut next_write_key = self.next_write_key.take();
-        let rows = flashback_to_version(
-            &mut txn,
-            &mut reader,
-            &mut next_lock_key,
-            &mut next_write_key,
-            self.key_locks,
-            self.key_old_writes,
-            self.start_ts,
-            self.commit_ts,
-        )?;
+        // The state must be `ScanLock` or `ScanWrite` here.
+        match self.state.as_mut().unwrap() {
+            FlashbackToVersionState::ScanLock {
+                next_lock_key,
+                key_locks,
+            } => flashback_to_version_lock(&mut txn, &mut reader, next_lock_key, key_locks)?,
+            FlashbackToVersionState::ScanWrite {
+                next_write_key,
+                key_old_writes,
+            } => flashback_to_version_write(
+                &mut txn,
+                &mut reader,
+                next_write_key,
+                key_old_writes,
+                self.start_ts,
+                self.commit_ts,
+            )?,
+        }
+        let rows = txn.modifies.len();
         let mut write_data = WriteData::from_modifies(txn.into_modifies());
         write_data.extra.for_flashback = true;
         Ok(WriteResult {
             ctx: self.ctx.clone(),
             to_be_write: write_data,
             rows,
-            pr: if next_lock_key.is_none() && next_write_key.is_none() {
-                ProcessResult::Res
-            } else {
+            pr: (move || {
```

Comment:
can't we just use `{ .. }` instead of closure?

Reply:
I think this is for injection testing to meet error in the first batch.🤔

Reply:
We need the failpoint return, so it seems that the closure is necessary.